src/main/scala/exu/dispatch-queue.scala

```diff
@@ -0,0 +1,449 @@
+    //  This double loop is a clonky way of checking it. It checks wether all of the valids at head
+    //  also match the ones dequeued this CC
+    val proceed = WireInit(true.B)
+    for (i <- 0 until enqWidth) {
+      val uop_done = WireInit(false.B)
+      for(j <- 0 until deqWidth) {
+        when(!valids(head)(i) || (io.heads(j).fire && deq_ptr(j) === i.U )) {
+          uop_done := true.B
+        }
+      }
+      when(!uop_done) {
+        proceed := false.B
+      }
+    }
+    head_next := Mux(proceed, WrapInc(head, numEntries), head)
+  }
+
+
+
+  // Handle branch resolution
+  //  On mispredict, find oldest that is killed and kill everyone younger than that
+  //  On resolved. Update all branch masks in paralell. Updates also invalid entries, for simplicity.
+  //
+  // This function takes a killed and valid Vec and finds out of the whole Vec was killed or is invalid
+  def width_killed(k: Vec[Bool], v: Vec[Bool]): Bool = {
+    v.reduce(_ || _ ) && (k zip v).map{case (kill, valid) => kill || !valid}.reduce(_&&_)
+  }
+  when(io.brinfo.valid) {
+    for (idx <- 0 until numEntries) {
+      for(lane <- 0 until enqWidth) {
+        val entry_match = valids(idx)(lane) && maskMatch(io.brinfo.mask, br_mask(idx)(lane))
+        when (entry_match && io.brinfo.mispredict) { // Mispredict
+          entry_killed(idx)(lane) := true.B
+          valids(idx)(lane) := false.B
+        }.elsewhen(entry_match && !io.brinfo.mispredict) { // Resolved
+          br_mask(idx)(lane) := (br_mask(idx)(lane) & ~io.brinfo.mask)
+          updated_brmask(idx)(lane) := true.B
+        }
+      }
+    }
+
+
+    // tail update logic
+    for (i <- 0 until numEntries) {
+      // treat it as a circular structure
+      val previous_killed =
+      if (i == 0) {
+        width_killed(entry_killed(numEntries - 1) ,valids(numEntries - 1))
+      } else {
+        width_killed(entry_killed(i-1) ,valids(i-1))
+      }
+      val this_killed = width_killed(entry_killed(i) ,valids(i))
+
+      // transition from not killed to killed - there should be one at maximum
+      when(!previous_killed && this_killed){
+        // this one was killed but the previous one not => this is tail
+        // if branches are killed there should be nothing being enqueued
+        // TODO: make sure this is true (assert)
+        tail_next := i.U
+      }
+    }
+  }
+
+  assert(! (io.brinfo.mispredict && io.brinfo.valid &&  width_killed(entry_killed(head), valids(head)) && (head_next =/= tail_next)), "[dis-q] branch resolution with head flushed but head and tail_next not reset")
+  assert(! (RegNext(io.brinfo.mispredict && io.brinfo.valid && width_killed(entry_killed(head), valids(head))) && (full =/= false.B || empty =/= true.B )), "[dis-q] branch resolution with head flushed but empty, full not reset")
+
+
+
+
+  require(enqWidth>=coreWidth)
+
+
+  for (w <- 0 until enqWidth)
+  {
+    io.enq_uops(w).ready := !full
```

Comment:
as soon as one valid is false it's not full
but that could be because head+1 was dequeued but not head

Reply:
I think it would work because:
- full = at least 1 valid per "row". If head+1 is dequeued and not head we would still have at least one valid in that row. 
- heads always belong to the same row. if head+1 is dequeued we can get (head, head+2) as heads but it will never run ahead to a new row.
- This means that the head row will count as full even when head+1 etc are dequeued